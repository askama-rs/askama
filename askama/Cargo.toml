[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
