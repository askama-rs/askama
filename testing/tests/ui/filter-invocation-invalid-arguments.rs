use askama::Template;

mod filters {
    #[askama::filter_fn]
    pub fn noargs(
        input: impl std::fmt::Display,
        _env: &dyn askama::Values,
    ) -> askama::Result<String> {
        Ok(format!("{input}"))
    }

    #[askama::filter_fn]
    pub fn req1(
        input: impl std::fmt::Display,
        _env: &dyn askama::Values,
        req1: usize,
    ) -> askama::Result<String> {
        Ok(format!("{input}{req1}"))
    }
}

#[derive(Template)]
#[template(source = r#"{{ "i like cake" | noargs(3) }}"#, ext = "html")]
struct InvokeNoArgFilterWithArgs;

#[derive(Template)]
#[template(source = r#"{{ "i like cake" | noargs(req1 = 3) }}"#, ext = "html")]
struct InvokeNoArgFilterWithNamedArg;

// ---------------------------------------------

#[derive(Template)]
#[template(source = r#"{{ "i like cake" | req1 }}"#, ext = "html")]
struct InvokeReq1FilterWithoutArgs;

#[derive(Template)]
#[template(source = r#"{{ "i like cake" | req1(3, "test") }}"#, ext = "html")]
struct InvokeReq1FilterWith2Args;

#[derive(Template)]
#[template(source = r#"{{ "i like cake" | req1(req2 = 42) }}"#, ext = "html")]
struct InvokeReq1FilterWithUnknownNamedArg;

// ---------------------------------------------

fn main() {
    InvokeNoArgFilterWithArgs.render().unwrap();
    InvokeNoArgFilterWithNamedArg.render().unwrap();
    InvokeReq1FilterWithoutArgs.render().unwrap();
    InvokeReq1FilterWith2Args.render().unwrap();
    InvokeReq1FilterWithUnknownNamedArg.render().unwrap();
}
