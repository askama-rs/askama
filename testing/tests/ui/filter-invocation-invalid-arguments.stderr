error[E0599]: no method named `with_0` found for struct `noargs<'filter>` in the current scope
  --> tests/ui/filter-invocation-invalid-arguments.rs:22:10
   |
 4 |     #[askama::filter_fn]
   |     -------------------- method `with_0` not found for this struct
...
22 | #[derive(Template)]
   |          ^^^^^^^^ method not found in `noargs<'_>`
   |
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: Argument at position 0 is invalid on filter noargs<'_>. Too many arguments supplied?
  --> tests/ui/filter-invocation-invalid-arguments.rs:22:10
   |
22 | #[derive(Template)]
   |          ^^^^^^^^ Filter function
   |
help: the trait `ValidArgIdx<0>` is not implemented for `noargs<'_>`
  --> tests/ui/filter-invocation-invalid-arguments.rs:4:5
   |
 4 |     #[askama::filter_fn]
   |     ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `Template` which comes from the expansion of the attribute macro `askama::filter_fn` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `with_req1` found for struct `noargs<'filter>` in the current scope
  --> tests/ui/filter-invocation-invalid-arguments.rs:26:10
   |
 4 |     #[askama::filter_fn]
   |     -------------------- method `with_req1` not found for this struct
...
26 | #[derive(Template)]
   |          ^^^^^^^^ method not found in `noargs<'_>`
   |
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: Argument at position 0 is invalid on filter noargs<'_>. Too many arguments supplied?
  --> tests/ui/filter-invocation-invalid-arguments.rs:26:10
   |
26 | #[derive(Template)]
   |          ^^^^^^^^ Filter function
   |
help: the trait `ValidArgIdx<0>` is not implemented for `noargs<'_>`
  --> tests/ui/filter-invocation-invalid-arguments.rs:4:5
   |
 4 |     #[askama::filter_fn]
   |     ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `Template` which comes from the expansion of the attribute macro `askama::filter_fn` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: Invalid filter function invocation. Not all required arguments were supplied.
  --> tests/ui/filter-invocation-invalid-arguments.rs:32:10
   |
32 | #[derive(Template)]
   |          ^^^^^^^^ Filter function
   |
help: the trait `ValidFilterInvocation` is not implemented for `req1<'_>`
  --> tests/ui/filter-invocation-invalid-arguments.rs:12:5
   |
12 |     #[askama::filter_fn]
   |     ^^^^^^^^^^^^^^^^^^^^
   = help: the trait `ValidFilterInvocation` is implemented for `req1<'_, true>`
   = note: this error originates in the derive macro `Template` which comes from the expansion of the attribute macro `askama::filter_fn` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `execute` found for struct `req1<'_>` in the current scope
  --> tests/ui/filter-invocation-invalid-arguments.rs:32:10
   |
12 |     #[askama::filter_fn]
   |     -------------------- method `execute` not found for this struct
...
32 | #[derive(Template)]
   |          ^^^^^^^^ method not found in `req1<'_>`
   |
   = note: the method was found for
           - `req1<'_, true>`
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `with_1` found for struct `req1<'filter, REQUIRED_ARG_FLAG_0>` in the current scope
  --> tests/ui/filter-invocation-invalid-arguments.rs:36:10
   |
12 |     #[askama::filter_fn]
   |     -------------------- method `with_1` not found for this struct
...
36 | #[derive(Template)]
   |          ^^^^^^^^
   |
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: Argument at position 1 is invalid on filter req1<'_>. Too many arguments supplied?
  --> tests/ui/filter-invocation-invalid-arguments.rs:36:10
   |
36 | #[derive(Template)]
   |          ^^^^^^^^ Filter function
   |
   = help: the trait `ValidArgIdx<1>` is not implemented for `req1<'_>`
           but trait `ValidArgIdx<0>` is implemented for it
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `with_req2` found for struct `req1<'filter, REQUIRED_ARG_FLAG_0>` in the current scope
  --> tests/ui/filter-invocation-invalid-arguments.rs:40:10
   |
12 |     #[askama::filter_fn]
   |     -------------------- method `with_req2` not found for this struct
...
40 | #[derive(Template)]
   |          ^^^^^^^^
   |
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)
